### ISSUE TITLE
Support or Migration Guide for Multiple Senders for Cloud Messaging

### ISSUE THREAD

**Issue by atakankaya** [firebase/firebase-android-sdk#4316]
## What feature would you like to see?

Previously, Cloud Messaging project supported the use of multiple app servers configured with different Firebase projects according to [documentation](https://firebase.google.com/docs/cloud-messaging/concept-options#receiving-messages-from-multiple-senders) (where this documentation is silently removed in October 2022 after the functionality was actually removed 1.5 year earlier at firebase-messaging:22.0.0). 

> Receiving messages from multiple senders
> FCM allows multiple parties to send messages to the same client app. For example, suppose the client app is an article aggregator with multiple contributors, and each of them should be able to send a message when they publish a new article. This message might contain a URL so that the client app can download the article. Instead of having to centralize all sending activity in one location, FCM gives you the ability to let each of these contributors send its own messages.
> 
> To enable this feature, make sure you have each sender's sender ID. When requesting registration, the client app fetches the token multiple times, each time with a different sender ID in audience field, using the token retrieval method for the given platform:
> 
> Swift — token(completion:)
> Android — FirebaseMessaging.getInstance().getToken()
> 
> Make sure that you do not add multiple sender IDs to a single token request, as this can have unpredictable results. Make each call separately, once per sender ID.
> 
> Finally, share the registration token with the corresponding senders, and they'll be able to send messages to the client app using their own authentication keys.
> 
> Note that there is limit of 100 multiple senders.

However, senderId cannot be passed to [FirebaseMessaging.getInstance().getToken()](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessaging#public-taskstring-gettoken).

Furthermore, firebase-iid dependency is removed in [firebase-messaging:22:0.0](https://firebase.google.com/support/release-notes/android#messaging_v22-0-0). According to the changelog: 
> Caution: This is a breaking change for apps that use FCM and the deprecated Firebase Instance ID API to manage registration tokens.
> We strongly recommend [migrating to FCM's token APIs](https://firebase.google.com/docs/projects/manage-installations#fid-iid). **If you're unable to migrate to the replacement APIs, add a direct dependency on the firebase-iid library to your build.gradle file.**

We defined `firebase-iid` as a dependency to continue using the deprecated [FirebaseInstanceId.getInstance().getToken(senderId, scope)](https://firebase.google.com/docs/reference/android/com/google/firebase/iid/FirebaseInstanceId#public-string-gettoken-string-senderid,-string-scope) as suggested in the release notes.
This seems to work for retrieving tokens for multiple senders; however, when tokens are refreshed;
- [FirebaseInstanceIdService.onTokenRefresh()](https://firebase.google.com/docs/reference/android/com/google/firebase/iid/FirebaseInstanceIdService#public-void-ontokenrefresh) is not getting called for any senders (tested with firebase-messaging:23.1.0 and firebase-iid:21.1.0). This is contrary to what the javadoc says:
> This method will be invoked on token changes even if onNewToken is also used.
-  [FirebaseMessagingService.onNewToken(token)](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessagingService#public-void-onnewtoken-string-token) is only getting called for the sender of the default project. It is not getting called for other senders. 

Because token refresh callbacks don't work and the deprecation due to migration from Firebase Instance ID to Firebase Installations ID, this doesn't seem to be a viable option. 

The second option is to initialize `FirebaseMessaging` using another `FirebaseApp` initialized with `FirebaseOptions` object for each other sender. But, there are multiple problems with this approach:
- [FirebaseMessaging.getInstance(FirebaseApp)](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-messaging/src/main/java/com/google/firebase/messaging/FirebaseMessaging.java#L156) is package-protected.
- [FirebaseApp.get(class)](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-common/src/main/java/com/google/firebase/FirebaseApp.java#L338) is marked with @KeepForSdk annotation.
- Even if we used `FirebaseApp.get(class)`; we have to pass ProjectName, ApplicationId, ApiKey to `FirebaseOptions` which are not known by us since the Firebase Projects for other senders are owned by other parties. Further, getting ApplicationId and ApiKey requires these third parties to set a new Android App in their Firebase Project for each of their consumers (Although our tests show that token retrieval and FCM messaging still works with our tester app using the ApplicationId and ApiKey belonging to a different package name!).
- [FirebaseMessagingService.onNewToken(token)](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessagingService#public-void-onnewtoken-string-token) is only getting called for the sender of the default project. It is not getting called for other senders. 

These were mentioned in different tickets (https://github.com/firebase/firebase-android-sdk/issues/3987, https://github.com/firebase/firebase-android-sdk/issues/3205, https://github.com/firebase/firebase-android-sdk/issues/2778, https://github.com/firebase/firebase-android-sdk/issues/2285, https://github.com/firebase/firebase-android-sdk/issues/4053) and official responses of the contributors of this project are contradictory (https://github.com/firebase/firebase-android-sdk/issues/3987#issuecomment-1230200424, https://github.com/firebase/firebase-android-sdk/issues/2778#issuecomment-952404884)

We would like to get clarity on the support for multiple senders feature.
We need this feature to be recovered; otherwise we would be stuck at firebase-messaging:17.0.0.

If this is not possible, **we would like to get a migration guide** since we are already live for 4 years with our multiple sender setup based on the removed docs since this would involve multiple parties to collaborate and make modifications in their apps.

## How would you use it?

Preferably, we would like to;
- have FirebaseMessaging.getInstance().getToken() API to be changed to FirebaseMessaging.getInstance().getToken(senderId)
- have FirebaseMessagingService.onNewToken(token) API to be changed to FirebaseMessagingService.onNewToken(token, senderId) and have it called for all sender IDs, not just the default one.

Our use case for multiple senders which is already live for 4 years is as below:
We have 3 senders for our app; 
- one owned by MasterCard (for MDES system) for mobile card tokenizations
- one owned by a third party company used with their Android SDK and their App Server
- one owned by our own App Server for other features of our app 

Removal of this feature and migration to single sender ID have huge implications for our company and our partners;
- we would have to share our private Web API Key of our App Server with other companies which is not acceptable from security perspective. Those other companies would use this API Key to send messages to our app. So, for example, MasterCard would need to get the private API Keys of the apps of ALL of their consumers. Each company would have to share their private API Key with MasterCard so that MDES can send FCM messages. But it would mean that they can send any FCM message, leaving us no way to verify if this FCM message is sent by our own App Server or MasterCard, therefore the security issue.  
- we would have to migrate every customer to use single token for single server. Since there is no guarantee that customer will launch our app immediately after update (and have connectivity to the internet); customers may not receive important FCM messages until each one are migrated. 
- both us and our partners have to make changes in their servers and their SDKs. 

Therefore, we prefer to have the functionality to be recovered; otherwise a clear migration guide for us and our partners.

**Comment by argzdev** [comment]
Hi @atakankaya, thank you for this well thought out request. Yes, I do believe that having a clear migration guide would be beneficial to the rest of our developers. Let me bring this up to our eng sync and see if we can get some feedback about this.

**Comment by atakankaya** [comment]
Hello @argzdev , 
Do you have any update on this ticket?

**Comment by argzdev** [comment]
Hi @atakankaya, sorry this is still being reviewed by our engineers. There's currently no update yet. We'll reply back here once we get further information. Thanks!

**Comment by argzdev** [comment]
Hi @atakankaya, thank you for your patience. 

I received some feedback from our engineers: We recommend using the V1 API for your use-case. In this way, the merchants own the FCM project and can either supply you with the credentials, or give your service account permissions to send FCM messages to merchants' project.

Here are some references that may be helpful in doing this approach:

- [Migrating from legacy to HTTP v1 API](https://firebase.google.com/docs/cloud-messaging/migrate-v1)
- [FCM roles](https://firebase.google.com/docs/projects/iam/roles-predefined-product#messaging) - they need to grant you the Admin role for the full read/write access to FCM resources.
- [Understanding service accounts](https://cloud.google.com/iam/docs/understanding-service-accounts) - service accounts can be used to call Firebase server APIs from your app server or trusted environment.
- [Authorize send requests](https://firebase.google.com/docs/cloud-messaging/auth-server) - with the v1 API, requests are authorized with a short-lived OAuth 2.0 access token instead of a server key. To mint this access token, you can manually obtain the required credentials from a JSON private key file generated for a service account.

Hope these resources help point you to the right direction. Feel free to reach back out if you have any questions or concerns.

**Comment by atakankaya** [comment]
@argzdev Could you elaborate on the word "merchant"? I am a bit confused there.

If I understood you correctly, you are proposing that; 
- We own one Firebase project for our Android app. 
- We create a Service account from our Firebase project and use its private key with our own App Server to send FCM messages to our Android app. 
- We create another Service account from our Firebase project, set its permissions and share its private key file with MasterCard.
- MasterCard's server keeps a map of consumer Android apps and service account private keys. (After all, we are not the only consumer Android app of their services)
- MasterCard's server uses the private key of the service account we shared and sends FCM messages to our Android app via v1 API.
- Each instance of our Android app generates one FCM registration token for our own Firebase project and sends it to both our own App Server and MasterCard's servers.

So, MasterCard should stop using their own Firebase project and handing out their SenderId to the consumer Android apps. 
Instead, each consumer of their services (so, each Android app owner) should create their own Firebase project and a service account for MasterCard and share its private key with MasterCard. 

Do I understand you correctly?

**Comment by atakankaya** [comment]
Hello @argzdev It's been 3 weeks, have you had a chance to check my questions with your developers?

**Comment by argzdev** [comment]
Hey @atakankaya, thanks for following up on us. For my previous response, sorry if it was unclear, I think I phrased it out incorrectly. The recommendation of our engineer is to use V1 API, where the merchants own the FCM project and supply MasterCard with credentials or give MasterCard's Service account permissions to send FCM messages to merchant's project.

> - We own one Firebase project for our Android app.
> - We create a Service account from our Firebase project and use its private key with our own App Server to send FCM messages to our Android app.
> - We create another Service account from our Firebase project, set its permissions and share its private key file with MasterCard.
> - MasterCard's server keeps a map of consumer Android apps and service account private keys. (After all, we are not the only consumer Android app of their services)
> - MasterCard's server uses the private key of the service account we shared and sends FCM messages to our Android app via v1 API.
> - Each instance of our Android app generates one FCM registration token for our own Firebase project and sends it to both our own App Server and MasterCard's servers.

Yes, I've also confirmed this with our engineers. You described each step perfectly, this is correct.

**Comment by atakankaya** [comment]
@argzdev Thanks for confirming. 

There are still 2 issues with your proposal:

1. I defined a service account with role: `Firebase Cloud Messaging Admin (roles/firebasenotifications.admin)` and tried to send an FCM message using v1 API. I am getting error: `...<removed>...Permission 'cloudmessaging.messages.create' denied on resource...<removed>...`
I can see [on the permissions docs you sent](https://firebase.google.com/docs/projects/iam/roles-predefined-product#messaging) that `Firebase Cloud Messaging Admin` role does not list this permission: `cloudmessaging.messages.create`

[On this page,](https://cloud.google.com/iam/docs/permissions-reference#search) I searched for that permission:
![permissions](https://user-images.githubusercontent.com/5691849/220156804-8681a124-36cf-475b-b2bb-a774d45cf241.png)
And I thought the only one that makes sense to use is `Firebase Cloud Messaging API Admin` (mind the `API`) role. 
But this role cannot be found on the IAM portal to add it to my test service account:
![IAM](https://user-images.githubusercontent.com/5691849/220157365-3eb804ab-5911-44c5-bafe-54330248a124.png)
 
Further, I found a [different page](https://cloud.google.com/iam/docs/understanding-roles#firebasecloudmessaging.admin) for list of roles and it is marked as Beta over there. 
![firebase roles](https://user-images.githubusercontent.com/5691849/220157756-1552aa41-f301-4d4c-b7cb-5f7af65bceaf.png)

Finally, I tested with `Firebase Admin` role, just to rule out any issue in my tester code; and it worked. But obviously, this is not a role I am going to be granting these service accounts.

So, it is still unclear to me which role to use because the ones you linked do not grant that permission. Can you clarify this please? 


2. How can the Android app determine which service account sent the FCM message?
I tested with 2 Service Accounts and both [RemoteMessage. getFrom()](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/RemoteMessage#getFrom()) and [RemoteMessage.getSenderId()](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/RemoteMessage#getSenderId()) are returning the same sender ID (which makes sense as there is 1 Firebase project) for the messages sent from both service accounts. 

But the question; how can we determine who sent the FCM message? Is it our own App Server or MasterCard? And this is important, because our own App Server sends custom "data messages" for which we show a push notification to the customer, whereas MasterCard sends custom encrypted "data messages" for some background processing which must not be displayed to the customer. How can I differentiate now which server sent the FCM message?

**Comment by argzdev** [comment]
Hi @atakankaya, given that the concern is no longer just related to the SDK, and your inquiries now require specific implementations, the level of expertise required may be better suited to our team of specialists. That said, could you file a ticket to our [Firebase support](https://firebase.google.com/support) team and reference this issue? They'll be able to assist you better and help you with your implementation requirements.

For others that have any more questions about the implementation, feel free to reach out to our support team. That said, I'll close this issue now. But if you think we missed something, just give us a reply and we'll take another look. Thanks!

**Comment by atakankaya** [comment]
Hi @argzdev,

After contacting Firebase support; we came to the conclusion that migration to "single project - multiple service accounts" setup is not feasible as this use case isn't well supported. Further, it is a very costly migration involving multiple parties.

However, I think you are missing below; 

We also contacted MasterCard and they pointed out that multiple firebase project setup is _still possible_. See these firebase docs on using multiple Firebase projects in the Android app: https://firebase.google.com/docs/projects/multiprojects#use_multiple_projects_in_your_application 

According to those docs, we can create a `FirebaseApp` instance for the other projects:

```
FirebaseOptions options = new FirebaseOptions.Builder()
     .setProjectId("<PROJECT_NUMBER>")
     .setApplicationId("<MOBILESDK_APP_ID>")
     .setApiKey("<API_KEY>")
     .build();
FirebaseApp app = FirebaseApp.initializeApp(context, options, "<UNIQUE_NAME_OF_APP>")
```

So therefore, MasterCard will provide the project number, mobilesdk_app_id and api_key (instead of SenderID).

Further, I see that multiple Firebase project is supported [from the docs of Firebase Installations SDK](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-installations/FCM_TOKENS_CHANGE.md);

> The latest Cloud Messaging SDK i.e firebase-messaging:20.1.1 updates introduce a transitive dependency on the [Firebase Installations SDK](https://console.cloud.google.com/apis/library/firebaseinstallations.googleapis.com), a new infrastructure service for Firebase. Different to before, the Firebase Installations SDK supports [multiple projects in your application](https://firebase.google.com/docs/projects/multiprojects).

And MasterCard pointed out that in order to retrieve FCM tokens for other firebase projects, we can simply call this;

```
String token = app.get(FirebaseMessaging.class).getToken();
```

I tested and it works perfectly. This way we can upgrade our Firebase Android SDK to the latest version without any migration needed on any backend systems.
 
However, there is a catch; [FirebaseApp.get()](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-common/src/main/java/com/google/firebase/FirebaseApp.java#L324) is annotated with `@KeepForSdk`. Why is that the case? And do you plan to remove this annotation?

**Comment by mkubiczek** [comment]
@atakankaya You can look into the issue we reported some time ago
https://github.com/firebase/firebase-android-sdk/issues/4450
There is some background regarding this API, which we also use for multiproject setup. And it also seem to work for us pretty well besides this odd issue described there.

**Comment by argzdev** [comment]
Hi @atakankaya, I've received updates from our engineers:

> @KeepForSDK marks methods in the SDK that are not part of the public API, and should only be used by other Firebase SDKs. There are no guarantees on how this functionality will be maintained; it can be kept or removed or changed as the need arises.

As for the migration setup, we've had some discussions and have taken a look into the conversation with the support team. And these are our recommendations:

1. Using a service account which grants countless permissions is causing a security risk.

> - Use custom scoped service account of only one permission: cloudmessaging.messages.create
> - Create a separate account for each vendor that is shared with, not one shared between all 3
This should mitigate the issue with the security risk. Even if one service account is compromised, changing that will not affect the other vendors.

2. Changing the message payload structure would mean a big complex migration

> - There isn't any other recourse, they will have to undergo some kind of migration where the message sender (as a third party) sends to their Firebase project with the service worker and oauth token, and include in the message data some indication of which sender it was. This will involve multiple parties, and require some updates to their Android client handling logic.

**Comment by mkubiczek** [comment]
@argzdev We are having similar use case as @atakankaya, where we were creating new Firebase Project for FCM when we had a need to add another push provider to our application. That way we had it nicely separated, and we could disable just one provider when needed without affecting other ones. Also using senderId to distinguish the push origin so it can be directed to proper handler. 

But then at some point you deprecated the FirebaseInstanceId, and at some point even removed from Firebase Messaging (although it's still available as separate library). We migrated to the new multi project setup, unfortunately using this @KeepForSdk API. Please note sometimes it's not possible to enforce once format of the push message and encode the senderid there (not mentioning the effort and number of parties involved). We cannot expect that the third party push providers will change the push structure for us. Also we would loose the flexibility we have now when using just single Firebase project.

I really think opening the API should be considered. I believe we are not the only ones having similar issues.

**Comment by atakankaya** [comment]
Thank you @argzdev. 

I noticed this in the[ changelog of firebase-messaging:22.0.0](https://firebase.google.com/support/release-notes/android#messaging_v22-0-0):
> ...If you're unable to migrate to the replacement APIs, add a direct dependency on the firebase-iid library to your build.gradle file...

Until the solution to this problem is clear; can we apply above solution? So, these steps: 
- Upgrade to `firebase-messaging:23.1.2`
- Manually add dependency to `firebase-iid:21.1.0`
- Continue using the API: `FirebaseInstanceId.getInstance().getToken(senderId, FirebaseMessaging.INSTANCE_ID_SCOPE)`

At this point, we are stuck at version 17.0.0 of `firebase-messaging` and a newer version of one of the third party SDKs we depend on in our app requires the latest version of the Firebase SDK. We couldn't upgrade that third party SDK since we were stuck. But if we are allowed to use above temporary solution, we could be unblocked. Is that OK?

**Comment by argzdev** [comment]
Hi @atakankaya, yes I've confirmed with our engineers. This is the recommended alternative solution for now, as long as the most recent versions of firebase-messaging and firebase-iid is used.

**Comment by atakankaya** [comment]
Hi @argzdev  

Looks like [FCM Legacy HTTP API will be decommissioned on 20 June 2024](https://firebase.google.com/support/faq#fcm-depr-features). So, I suppose this alternative solution (with `firebase-iid`) won't work any more. 

> That said, your worries are justified. We'll discuss this more in our engineering syncs and see if we can come up with better alternative solutions, although I must emphasize that there is no guarantee that a better option will be found.

Have you discussed this and do we have a better solution now? 

Especially, around the issue with missing sender information. 

If Android clients cannot know which server sent the message, then our message payload has to undergo some changes to reflect the sender in them. This means, not only will we have to migrate the message payloads sent from the servers our app subscribed to, but also it indirectly means we have to align our message payload with all MasterCard's consumers (which we have no direct relation to). Honestly, I don't find this realistic...

**Comment by argzdev** [comment]
Thanks for notifying us again, @atakankaya. 

Have you discussed this and do we have a better solution now?
> Sad to say, there wasn't much of updates.

Yeah that is reasonable, let me bump this up again since the decommission without mitigation is a total cut off of support, and as you said the transition change doesn't sound too feasible.

**Comment by atakankaya** [comment]
Thanks @argzdev ,

I made an assumption there though, maybe I was wrong; 

Would it still be possible to use our alternative approach with `firebase-iid` after migrating the backends to use V1 API? 
So, would calling the deprecated `getToken` of `firebase-iid` with sender ID still work after 20 June 2024? 
If so, then we could still stick to the current solution I think (while migrating the backends to V1 API).

**Comment by argzdev** [comment]
Hi @atakankaya, I've just got update from our engineers. `FirebaseInstanceId.getInstance().getToken(senderId, FirebaseMessaging.INSTANCE_ID_SCOPE)` will still continue working and will not be affected by the FCM Legacy HTTP API decommission, so you would only need to do the migration to v1 API with your backends.

**Comment by mkubiczek** [comment]
Hi Firebase team, wonder if you did break the multi project setup with the introduction of the [Firebase Instance ID based FCM API](https://github.com/firebase/firebase-android-sdk/pull/8087)?

We are having multiple Firebase FCM projects within the app, but looks like with the recent change it's going to stop working.
In our setup we instantiate multiple FirebaseMessaging instances. Then we utilize the `FirebaseMessgaing.getToken`, `FirebaseMessgaing.deleteToken`. Despite the old `onNewToken` only called for the default Firebase FCM project, we could still rely on the mentioned APIs, to generate or delete the tokens for the other projects.

In the new API the `FirebaseMessaging.register` method doesn't return the token anymore, but I assume on successful registration the regular firebase installation id can be used.
The problem is, with previous approach we could read sender id from the remote message and use it to pass that message to appropriate handler, now you suggest using FirebaseOptions.gcmSenderId instead, but how do I get proper FirebaseOptions instance, there should be something in the remote message we could rely on, like keep the sender id there, or pass there something else like firebase project id or firebase installation id.
